src/Lucene.Net/Util/UnmanagedStringArray.cs

```diff
@@ -0,0 +1,185 @@
+                    return 0;
+
+                if (strB.IsNull)
+                    return 1;
+
+                if (strA.IsNull)
+                    return -1;
+
+                return strA.StringAsBytes.SequenceCompareTo(strB.StringAsBytes);
+            }
+
+            public static int CompareOrdinal(UnmanagedString strA, string strB)
+            {
+                if (strA.IsNull && strB == null)
+                    return 0;
+
+                if (strB == null)
+                    return 1;
+
+                if (strA.IsNull)
+                    return -1;
+
+                Span<byte> managed = Encoding.UTF8.GetBytes(strB);
+                return strA.StringAsBytes.SequenceCompareTo(managed);
+            }
+
+            public static int CompareOrdinal(string strA, UnmanagedString strB)
+            {
+                return -CompareOrdinal(strB, strA);
+            }
+
+            public int CompareTo(object other)
+            {
+                return CompareOrdinal(this, (UnmanagedString) other);
+            }
+        }
+
+        private UnmanagedString[] _strings;
+        private List<Segment> _segments = new List<Segment>();
+        private SortedDictionary<long, Segment> _sortedSegments = new SortedDictionary<long, Segment>();
+
+        public int Length => _index;
+        public int _index = 1;
+
+        public UnmanagedStringArray(int size)
+        {
+            _strings = new UnmanagedString[size];
+        }
+
+        private Segment GetSegment(int size)
+        {
+            if (_segments.Count == 0)
+                return GetAndAddNewSegment(4096);
+
+            // naive but simple
+            var seg = _segments[_segments.Count - 1];
+            if (seg.Free > size)
+                return seg;
+
+            return GetAndAddNewSegment(Math.Min(1024 * 1024, seg.Size * 2));
+        }
+
+        private Segment GetAndAddNewSegment(int segmentSize)
+        {
+            var newSegment = new Segment(segmentSize);
+            _segments.Add(newSegment);
+            _sortedSegments.Add(newSegment.Number, newSegment);
+            return newSegment;
+        }
+
+        public void Add(Span<char> str)
+        {
+            fixed (char* s = str)
+            {
+                var size = (ushort) Encoding.UTF8.GetByteCount(s, str.Length);
+                var segment = GetSegment(size + sizeof(ushort));
+
+                segment.Add(size, out var position);
+                fixed (byte* r = new Span<byte>(position + sizeof(ushort), size))
+                {
+                    Encoding.UTF8.GetBytes(s, str.Length, r, size);
+                }
+                _strings[_index].Start = position;
+                _strings[_index].Owner = segment.Number;
+                _index++;
+            }
+        }
+
+        public UnmanagedString this[int position] => _strings[position];
+
+        public void CopyTo(UnmanagedStringArray dest, int destPosition, int position)
```

Comment:
I don't follow why we need this and the `_sortedSegments` at all.
You are using to copy a value, but the lifetime of the `lookup` is the same as the `values`, so shouldn't be an issue there.